x/mongo/driver/operation.go

```diff
@@ -321,8 +321,74 @@ func (op Operation) shouldEncrypt() bool {
 	return op.Crypt != nil && !op.Crypt.BypassAutoEncryption()
 }
 
+// filterDeprioritizedServers will filter out the server candidates that have
+// been deprioritized by the operation due to failure.
+//
+// The server selector should try to select a server that is not in the
+// deprioritization list. However, if this is not possible (e.g. there are no
+// other healthy servers in the cluster), the selector may return a
+// deprioritized server.
+func filterDeprioritizedServers(candidates, deprioritized []description.Server) []description.Server {
+	if len(deprioritized) == 0 {
+		return candidates
+	}
+
+	dpaSet := make(map[address.Address]*description.Server)
+	for i, srv := range deprioritized {
+		dpaSet[srv.Addr] = &deprioritized[i]
+	}
+
+	allowedIndexes := make([]int, 0, len(candidates))
+
+	// Iterate over the candidates and append them to the allowdIndexes slice if
+	// they are not in the deprioritizedServers list.
+	for i, candidate := range candidates {
+		if srv := dpaSet[candidate.Addr]; srv == nil || !srv.Equal(candidate) {
```

Comment:
Optional: How do you think using `if srv, ok := dpaSet[candidate.Addr]; ok || !srv.Equal(candidate) {`

Reply:
Good idea!